woocommerce/woocommerce-ios
# [Mobile Payments] Pair a reader, Mark I: CardReaderService

Closes #3714 

⚠️  This PR is against the feature branch implementing support for Card Present Payments, not against develop ⚠️ 

We continue building our prototype (for reference, p91TBi-4q4-p2), this time, by adding enough code to the CardReaderService to attack the Stripe Terminal SDK (in test mode) and request a connection to a given reader.

Testing this is very tricky. There are indications for settings up and running the integration tests (that cover this use case) in the diff of the README.md that accompanies the PR. But, to be honest, the easiest way to test this will be when #3715  is implemented, which is something I plan on doing right away.

## Changes
* Add enough code to StripeCardReaderService to pair with a card reader. Because we do not want to expose the model objects provided by the Stripe Terminal SDK to the rest of the codebase, we cache the discovered readers, of type `StripeTerminal.Reader`, until the connection process is complete. [The documentation suggests this is possible](https://stripe.dev/stripe-terminal-ios/docs/Protocols/SCPDiscoveryDelegate.html#/c:objc(pl)SCPDiscoveryDelegate(im)terminal:didUpdateDiscoveredReaders:):
> You should not cache the SCPReader objects in the readers array. In other words, after calling connectReader: with a discovered reader, you should not attempt to cache and re-use the Reader object

* Added a skeleton of an enumeration to model errors. This will be fully implemented in #3734 
* An integration test to trigger discovery, then trigger a connection to the first discovered reader, and then observes if the `connectedReader` property of CardReaderTerminal is updated. Instructions for setting up to run this test are available in the readme file that is part of the diff. If the test is setup successfully, the simulator will ask for permission to access Bluetooth. There is no way to recover, at this point, if the request is not granted. We'll get there though...

## How to test
This one is tricky. There are instructions in the diff to setup a testing site and obtain a connection token to run the integration test. 

* Checkout the branch
* run `bundle exec pod install`
* Activate StripeCardReaderIntegrationTests
* Run StripeCardReaderIntegrationTests (see instructions in stripe-tests.md)

Update release notes:
